packages/block-editor/src/components/block-list/block-wrapper.js

```diff
@@ -265,6 +269,50 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
 		};
 	}, [ isNavigationMode, isHovered, setHovered ] );
 
+	useEffect( () => {
+		// To accurately determine which blocks are hovered we must look at the elements under the cursor.
+		// Internal block inserters will fire events like mouseleave for blocks they are visually contained within,
+		// which makes an add/remove by clientId per individual block approach inaccurate.
+		function getHoveredBlocksFromCursor( event ) {
+			const hoveredElements = document.elementsFromPoint(
+				event.clientX,
+				event.clientY
+			);
+			const blockIds = [];
+			hoveredElements.forEach( ( element ) => {
+				if ( element.dataset.block ) {
+					blockIds.push( element.dataset.block );
+				}
+			} );
+			return blockIds;
+		}
+		function evaluateHoveredBlocks( event ) {
+			// Check the timeStamp of the last time this was set.
+			// This prevents needlessly rerunning the parse and dispatch for nested blocks
+			// who share boundaries that trigger the mouse events at the same time.
+			if ( event.timeStamp !== getHoveredBlocksTimeStamp() ) {
```

Comment:
Would it make sense to store this timestamp locally (maybe with `useRef`?) instead of having to send it to the store?

Reply:
Wouldn't that still create a different ref for each instance of the wrapper?

Reply:
Id like to keep it out of the store as well, but I wasn't quite sure how to make that work. 🤔

Reply:
> Wouldn't that still create a different ref for each instance of the wrapper

ahh good point. I guess it would have to be at a higher level, or in the global file scope. Global file scope would be really easy to do, but it doesn't feel "best practice" I guess

Reply:
🤔   Yeah... Im not sure which way is better, I'm open to either direction.  While I am not a huge fan of complicating the store with it, it is optional and doesn't necessarily need to be used by anyone else ever setting the values.

Reply:
While I'm also wary of adding this to the store, I see this as a legitimate use case. The only other alternatives I can think of are using context or using local state in a common parent component and prop drilling (`packages/edit-site/src/components/block-editor/index.js` maybe?)

I'd also prefer to avoid global file scope for state if only because it subverts expectations and its another pattern to keep track of when handling state management for components.

Reply:
I wondered about prop drilling as well, it seems like it may be a bit more ugly.  There may be a better way than I am thinking, but it would need to be passed into `useBlockWrapperProps` everywhere it is used (which includes a lot of edit.js block files).  I think the store may be the cleanest home for the timestamp.

> While I'm also wary of adding this to the store, I see the timestamp as a legitimate use case. It's information that needs to be shared amongst many components.

Very true.